### ISSUE TITLE
Project related news (was: Gathering interested people)

### ISSUE THREAD

**Issue by falkTX** [DISTRHO/Cardinal#1]
Leave a comment if you want to help.
This does not do anything yet besides building VCV sources, I need to do some experiments for it first to see what is the best approach.
Seems that we only need to create a variant of https://github.com/VCVRack/Rack/blob/v2/adapters/standalone.cpp for plugins, and then add in some hardcoded module for IO (audio, midi and transport sync)

**Comment by falkTX** [comment]
Update: the makefile builds now.
You get a black screen if you try to run it, as there is no real code on the plugin side yet.

**Comment by unfa** [comment]
Being able to use VCV Rack as an LV2 plug-in would put it on my map. I could actually start using it in my work.
It'd be something very unique, especially if it'd allow putting external audio in.

**Comment by falkTX** [comment]
CVCRack reads as "CV Crack", for those addicts on CV. Hopefully everyone finds it funny and not problematic :sweat_smile: 

One initial question is knowing which plugin IO variants are useful, because not all hosts are alike and allow placing FX on sequence tracks or synths on the mixer.
So I am thinking:

1. base synth version - no audio ins, MIDI input, 2 audio outs
2. base FX version - 2 audio ins, 2 audio outs, midi input and output
3. CV port version (only possible in LV2 and VST3, not in VST2) maybe with like 8 cv ins and outs, opinions welcome
4. ???

At least for FL Studio and Renoise we need a basic synth and fx versions.
The CV version wont work on most hosts, but useful for placing it on Carla and other modular things. I dont think JUCE hosting supports CV, so no support for Bespoke :/

**Comment by falkTX** [comment]
For the sake of transparency, I am quoting contents of an email I just received from Andrew:

> Hello,
> The name you've chosen for your fork is very close to our trademark "VCV" (USPTO 87714461) which may lead to confusion. VCV Rack's own 250k+ users are extremely brutal/demanding for tech support, and the last thing > I need is for a similar-looking product with a similar name creating more support due to confused users. I already get hundreds of complaints per year about other forks of VCV Rack (having much more different names than "CVC") and have to send them an automated template response explaining that we don't make that product and to contact who they obtained their version from. This makes many people angry toward VCV for not offering them the free support they think they deserve!
> 
> VCV releases open-source software so people can review and modify how the software works. We don't release software because we like providing support for free. I hope you can understand this and change the name of your fork to something that will cause fewer problems. Perhaps "falk Rack" would be better, as it suggests who users will assume is responsible for support?

And my reply:

> Hi Andrew, that makes sense, sure.
> 
> Though you could have said the same thing on the github ticket where conversation is on-going, I dislike to keep such matters private.
> 
> It is only a research project at the moment, if anything comes of it I will rename it (likely even before that).

**Comment by falkTX** [comment]
A few more details, seem that passing events from DPF to VCV works well.
Have a hacky way to pass-through mouse clicks, motion/hover and scroll.
With a little resize handler at the bottom right the UI can become resizable even in VST2 format.
Seeing in action as below: (Renoise and VST2 version)

<img width="1280" alt="image" src="https://user-images.githubusercontent.com/1334853/136480460-196d01d9-bb37-457a-95ce-6377aba756d8.png">

The SVGs are not loading somehow, dont know why yet.

The way VCV manages the "active context" is a bit weird for plugin usage, but should still be doable.
oui-blendish kinda works the same way and my testing for the one-knob series tells me it is not an issue as long as host GUI is single-threaded (which so far is always the case)

Since VeeSeeVSTRack had a working audio plugin engine, I assume this part is doable too.
We can likely copy its plugin modules to serve as a base.

The hard part is going to be dealing with VCV plugins, VeeSeeVSTRack went a bit overboard with its approach even though long-term it is/was the more stable one. Audio plugins should be self-contained, in theory, but in practice VCV already fails due to its heavy use of global state. So perhaps not worth bothering too too much on this side of things...?

Oh and yes this needs a new name. I am often bad at those, so suggestions welcome.

**Comment by x42** [comment]
> Oh and yes this needs a new name. I am often bad at those, so suggestions welcome.

How about "Cardinal"?

**Comment by falkTX** [comment]
> > Oh and yes this needs a new name. I am often bad at those, so suggestions welcome.
> 
> How about "Cardinal"?

I am quite neutral about this one, what is the reason for it?
Starts with a "C"...?

**Comment by falkTX** [comment]
I dont see why not, thanks. It actually sounds cool but still generic and simple enough.
We just got to see if there are no other plugins or big projects already with the same name

And not all projects start with C, you know DISTRHO itself is already an exception right? 😉

**Comment by falkTX** [comment]
The embed view works on Linux/X11, and tested win32 builds via wine as well.

![Screenshot_20211008_154900](https://user-images.githubusercontent.com/1334853/136578170-d8b178d4-e72b-423d-a894-9fab39286011.png)


If we can figure out why svgs dont load correctly, I think then a cleanup for a proper UI can begin.
The DSP I am assuming can just work, just a matter of setting up the engine properly.
I was more doubtful of the UI because of GLFW, still am a bit because when using stock Window.cpp that uses GLFW, the svgs then load correctly:
![Screenshot_20211008_160325](https://user-images.githubusercontent.com/1334853/136579875-46132e2e-6de3-4899-abf1-b24fbfc3750a.png)

My guess would be at some OpenGL setup/state not being quite as needed when using DPF/pugl

**Comment by falkTX** [comment]
Multiple instance seems to work, at least for UI stuff.
Some barebones keyboard input too, still need to handle special keys so we can do stuff like backspace for deleting characters.

<img width="1280" alt="image" src="https://user-images.githubusercontent.com/1334853/136656352-430678b5-851b-46f9-a0bf-f43bdda336d2.png">

This is showing how great it is to build good infrastructure.
Most of the work is just glueing stuff together, tying the DPF things into what VCV expects. It goes rather smoothly, so fingers crossed...

The next big item I guess is getting the engine to do something, likely I will need help on this.
But for that the build needs to work cleanly, so others can start to run it too and not just me.
Since we have CI tests already in place from other DPF plugins it should be clear when the build and plugin runtime is working well enough. This might take some time now though...

**Comment by Be-ing** [comment]
Is or will there be a means for an LV2 host to tell the plugin to load a particular patch separate from other instances? My hope is that I can save patches and create effect chain presets in Mixxx which load specific patches, then manipulate them with a single knob on my controller:
![image](https://user-images.githubusercontent.com/9455094/136658214-3e5b3788-9f8e-4be7-b77f-a3a14bf7e96b.png)

**Comment by falkTX** [comment]
> Is or will there be a means for an LV2 host to tell the plugin to load a particular patch separate from other instances?

Yes of course, each instance of Cardinal is supposed to be self-contained in its own engine, patch and gui.
And state should be preserved in the host as long as the host saves it and restores it the usual way.
For LV2 and Mixxx just make sure to support the LV2 state extension, 



> Are modules' knobs exposed as LV2 parameters?

This is up to discussion.

> IIUC LV2 plugins can't dynamically add and remove parameters, so I suppose it would need a hack similar to audio I/O with a fixed number of LV2 parameters which could bind to modules' knobs within the plugin.

That is correct. Though VST3 supports this dynamic parameter stuff, it is not something I am interested on adding to DPF.
The complexity around it is too much for very little gain, and almost no plugins need this in practice.

Hosts dont typically support CV as a form of automation, so we will need some special parameter mapping.
What I have done in Carla (and zynaddsubfx for example does something similar) is to expose a static 100 parameters to the host. In Carla these will be auto-assigned to the first plugins loaded in the rack, in zynaddsubfx you can map as you wish through a dedicated dialog.
For VCV as plugin I think the static parameters are the best choice. We can have a dedicated module to convert from plugin parameters to VCV internal CV, probably with some options to control its behaviour (sudden changes vs smooth interpolation, threshold for smoothness etc)

Nothing about this is done yet, so for now it is just discussion.
Opinions always welcome, specially before any real implementation begins.

**Comment by Be-ing** [comment]
> It would be nicer to have some fixed "macro" controls. Say, 24. They could show up in a dedicated "Control" module as CV source to connect wires to. The same module could also provide host-clock, MIDI etc.

This seems like a good idea and likely more practical than exposing every knob of every module.

**Comment by dromer** [comment]
@AndrewBelt if you would use DPF you could have vst2/3 without steinberg (and juce?) commercial license restrictions. ISC would give you the freedom to develop your vcv products your way.

**Comment by falkTX** [comment]
VST2 without steinberg and closed-source is not possible. see https://github.com/DISTRHO/DPF/blob/develop/LICENSING.md
VST3 can though, because it is a full API reimplementation, no SDK parts are used.

**Comment by dromer** [comment]
yes noted, but @AndrewBelt seems to have a vst2 license already, so potentially any distribution under that format would be covered by his license?
vst2 could be default off. only for the commercial vcv builds applied

**Comment by dromer** [comment]
But even if steinberg where to claim that vestige is bound by vst2 licensing a valid vst2 license could allow commercial use by vcv.

**Comment by AndrewBelt** [comment]
Have no worries. We are thousands of hours and tens of thousands of $ into developing and testing the plugin version of VCV Rack on 20 DAWs on 3 OS's while maintaining stability and compatibility with all plugins and patches. We are on-target for a November release of VCV Rack 2 Community and Studio Edition, and pre-release versions will be available for plugin developers in a few days.

**Comment by falkTX** [comment]
So that means a plugin version is known to work well, good stuff :)

Some news on the build setup, finally all the CI stuff passes. From linux to macos to windows (via mingw)
See https://github.com/DISTRHO/Cardinal/actions/runs/1324743342

macOS was the trickiest one to deal with, but we have x64+arm64 universal builds.
the linux stuff includes armhf, aarch64, x86 and x86_64.

These automated builds are not useful yet, because I do not have in place a way to detect the plugin bundle folder (we need something that works for all formats, discussion follows on https://github.com/DISTRHO/DPF/issues/332 for this)

For now at least you should be able to build and run it. The VCV "system" directory is just going to be to build-time source code directory.
Still need to do a couple more things and start the tests regarding the engine and multi-threading, but we are not too far off from something where I can distribute some tasks to those that want to help.

**Comment by Be-ing** [comment]
The way you are structuring this repository using upstream Rack as a Git submodule is interesting. I was thinking you'd have to modify the Rack code directly, but I suppose this approach will be easier to maintain.

I am interested to get this working in Mixxx but there is a lot of work to do in Mixxx first, so I probably won't be helping much with the plugin code before that.

**Comment by falkTX** [comment]
> The way you are structuring this repository using upstream Rack as a Git submodule is interesting. I was thinking you'd have to modify the Rack code directly, but I suppose this approach will be easier to maintain.

So far I have not have the need to edit header files, so it still works.
For C++ files we can just build/ship our custom implementation, as it is the case with the Window.cpp already for the integration with DPF UI stuff.

Still not optimal IMO, some code has to be duplicated in those headers (I mean, the same code is present on VCV and we only need to modify parts of it).

I wonder about the file menus we need to hide, in theory should be possible to access them through the Window methods, going deeper into the widgets until it reaches the menu items and then call ->hide() on them.
For example the file->quit is one that doesnt work if the UI is embed, needs to be hidden away then


> there is a lot of work to do in Mixxx first, so I probably won't be helping much with the plugin code before that.

Just make sure LV2 plugins work as good as possible, then you get these and many others for free :)

**Comment by falkTX** [comment]
I got distracted with a little new test project https://github.com/DISTRHO/Ildaeil/ as someone is adding (REAPER's) JSFX plugin support in Carla and I wanted it to be usable for other hosts that are not Carla. It served as motivation for a super stripped-down mini-host based on Carla.

Anyway, continuing with the updates...

I am skipping the engine + plugin process sync a bit, and went directly to the more challenging bit - plugins.
The engine provides its own sets of challenges, but it more of understanding and studying the VCV code rather than overcoming a challenge.

So I began trying to add plugin support.
The build setup was getting quite messy with the plugins inside VCV as plugin, so I just reorganized the whole thing.
Now it is more clearly separated between "deps", "plugins" (ie, vcv based plugins included as submodules) and "src" for the main VCV as plugin stuff.

Added AudibleInstruments, Befaco and Fundamental so far.
The AudibleInstruments is more complex than the other 2, due it needing many more source files.
Checkout https://github.com/DISTRHO/Cardinal/commit/c1233ab66e4fc6fb6e4ea3bc6e7c7b5dd4d5b748 for get an idea of how much work (or not) it is to add new vcv plugins to the build.
TL;DR we basically add a repo as submodule and adjust the Makefile and plugins.cpp files based on the original code.

And before you ask, it is all built statically into the Cardinal plugin binary.
With these 3 plugin sets the optimized binary is at 6.8Mb.

And to keep up with the tradition, here is yet another screenshot. Cardinal inside Carla in macOS, loading a few vcv plugins.

<img width="1280" alt="image" src="https://user-images.githubusercontent.com/1334853/137605675-2fea6f55-2312-4499-a665-6d068e5d3619.png">

**Comment by Be-ing** [comment]
> And before you ask, it is all built statically into the Cardinal plugin binary.

Is this a technical requirement? Is it feasible to load other Rack plugins dynamically?

**Comment by Be-ing** [comment]
Okay, seems reasonable to just statically link all the Rack plugins you can, especially considering how small the binary is.

**Comment by zezic** [comment]
Got the 5efc37e15caa06bb3b8feaa90b6af9d7ea192fc9 working on Linux, but it has a large empty black area for some reason before actual background starts: 
![image](https://user-images.githubusercontent.com/1945127/137637991-580e7901-7b96-4520-ac66-a6e7d1a26d53.png)
I can pan the area with BG into view with middle mouse button though.

**Comment by zezic** [comment]
Just built and checked the latest DPF plugins and they are all rendered correctly. @falkTX as you can see, the top (blendish) menubar of Rack is placed where it should be and only the main workspace has that strange offset of black color.

**Comment by Be-ing** [comment]
Awesome work!

**Comment by zezic** [comment]
Regarding modules popularity: https://colab.research.google.com/drive/1Rniemjz4xDzxaTrTEXdFjwhVLhptxloD#scrollTo=413f0952&line=1&uniqifier=1

**Comment by falkTX** [comment]
> Regarding modules popularity: https://colab.research.google.com/drive/1Rniemjz4xDzxaTrTEXdFjwhVLhptxloD#scrollTo=413f0952&line=1&uniqifier=1

That is super useful, thanks!

**Comment by zezic** [comment]
@falkTX Thanks for the update, the black offset problem on HiDPI is solved, but now the mouse coordinates are wrong: https://youtu.be/LFACL-FIkto
They were fine before the fix.

**Comment by zezic** [comment]
>Right, I know why! I was testing some things and forgot to add back the scale factor divider.
Should be fixed now again, though I didnt test this yet.

Thanks, I just tested it and it works correctly now!

**Comment by fpesari** [comment]
> On the CC-ND (non-derivatives) we are not changing the files, so not sure how that would even apply.

I was more worried about packagers not being able to be include Cardinal in distros which do not allow nonfree stuff (VCV Rack has the same problem).

**Comment by falkTX** [comment]
Ok yeah for the Non-Commercial clause the situation is a bit more tricky..

**Comment by Be-ing** [comment]
Could you make a build option to exclude CC-NC content so distros could package this who can't include it?

**Comment by falkTX** [comment]
Plugin state save/restore is in, as well as 24 host-exposed parameters.
The (added by default) "host parameters" module will map between host parameters and the VCV world.
Some little filtering to reduce stepping and it produces nice results.

I was able to get Carla playing a MIDI file where the notes are converted into parameter changes that automate Cardinal's parameters.
The result is (sorta) playing the chocobo song purely out of plugin parameter automation. :tada: 

https://user-images.githubusercontent.com/1334853/138166313-9757f351-abb9-4993-94d9-e34082fc5f47.mp4

**Comment by falkTX** [comment]
I have some quite bad news.
After looking around to perhaps replace the Core or ComponentLibrary with free alternatives, pretty much all projects that make VCV modules are also encumbered by the same very-restrictive license.

These are CC BY-NC-SA 4.0
https://github.com/bogaudio/BogaudioModules/blob/master/LICENSE.md

These are CC BY-NC-ND 4.0
https://github.com/MarcBoule/ImpromptuModular/blob/v2/LICENSE.md
https://github.com/MarcBoule/MindMeldModular/blob/v2/LICENSE.md

Even parts of the modules made by @AriaSalvatrice can't escape this, quoting from the project README:

> The Arcane module uses Tarot cards altered from Yoav Ben-Dov's CBD Tarot. Its graphics are distributed under the CC-BY-NC-SA, to comply with the license of the graphics used. That includes the faceplate, as it uses a pattern taken from that tarot deck.

So this is a very big mess for packaging, the NC clause basically prevents it to going in any linux distribution repository.

We have 2 options here:
1. Make Cardinal a very restrictive product with only 100% free modules, while being able to be packaged, sold etc without restrictions. The usefulness of such a product is questionable due to how few 3rd party modules we can import
2. Keep modules regardless of license, which will make Cardinal unpackageable and restricted on commercial things, but on the other hand we get some of the modules people are used to

**Comment by fpesari** [comment]
Well as a packager I am sorry but if this project can help someone, I say 2)...maybe outside of DISTRHO, since the official website calls DISTRHO "an open-source project for Cross-Platform Audio Plugins." and this is not open-source. 

But, as a purely personal opinion of someone who has followed your work for years, I think your limited time and expertise is better spent on projects which benefit the wider FLOSS audio community.

**Comment by falkTX** [comment]
It is a long shot but I have asked bogaudio developer to lift the NC restriction, at https://github.com/bogaudio/BogaudioModules/issues/187
A similar thing in https://github.com/MarcBoule/ImpromptuModular/issues/42

If one of them remove the NC clause I might go for the option 1 above.

**Comment by falkTX** [comment]
> Well as a packager I am sorry but if this project can help someone, I say 2)...maybe outside of DISTRHO, since the official website calls DISTRHO "an open-source project for Cross-Platform Audio Plugins." and this is not open-source.

Right, if we end up on that path then moving Cardinal out of DISTRHO would make sense.
Also to keep around any necessary module forks without polluting the DISTRHO org repos.



> But, as a purely personal opinion of someone who has followed your work for years, I think your limited time and expertise is better spent on projects which benefit the wider FLOSS audio community.

This project has helped to improve DPF quite a lot actually, since it pretty much uses all of its features.
Specially for file dialogs and text input which are barely used in the current plugins.

Plus I already started tests to have individual VCV modules are their own "standalone"/self-contained LV2 plugins.
There are quite a few modules that would be great to have in modular host environments like Carla, Ingen and SynthPod.

But most of all, working on Cardinal is fun :)

**Comment by falkTX** [comment]
There may be a way for the license deadlock...

To explain and go back a bit, the main dilemma is that some VCV files are CC-ND (non-derivatives) and VCV readme states that VCV logo and name cannot be used in derivative works.
The logo is used on CC-ND artwork, so by removing the logo we break the ND clause.

What if we would **not** modify the files but instead programmatically hide the logo? The files stay intact and no logo or VCV name is shown.
We are also not distributing the logo in the source code, we simply link to the Rack repository as a submodule.
So no derivative work is ever published, the maximum we can say is that one is being generated on the fly on the user's computer.

I did some quick tests, we can even delete the VCV name from the loaded shapes in memory so it is not just hidden but truly gone.
Screenshot of the default template without VCV logos:

<img width="1280" alt="image" src="https://user-images.githubusercontent.com/1334853/140431490-ee5cc552-8174-449f-b67b-3bbe8c976b67.png">

@AndrewBelt is this something you would be ok with?
No "VCV" is shown to the user but at the same time no Core/ComponentLibrary files are modified either.

**Comment by falkTX** [comment]
Some more testing for OpenGL things..
Since we can draw using regular GL stuff, bringing imgui was a first step, and then afterwards was [Ildaeil](https://github.com/DISTRHO/Ildaeil).

Still need to clean this up after the tests and finalize keyboard input, but seems to work quite ok.
We have our own plugin host inside Cardinal now :rocket: 

![Screenshot_20211105_225753](https://user-images.githubusercontent.com/1334853/140587269-24b4c0d6-723f-4bce-bf9f-35f69555aef3.png)

I did not bother with embedding the UI, sounds like it would be super complicated.
Alike the VCV commercial plugin host, there is an amount of latency using such plugins. I set it fixed at 128 frames for now.
There is no MIDI, as I dont really know how that could even work.. transport could be done by just mapping the host provided time info to Ildaeil.
And only LV2 hosting for now, since that is what Ildaeil does too. But since this is all done through [Carla](https://github.com/falkTX/Carla) we can very likely just have dedicated Carla modules in there later on. Will be specially useful for its Patchbay+CV mode.

**Comment by falkTX** [comment]
Everyone likes dark mode right?
Inverting the colors (programmatically again, so we do not modify artwork files) we get what is in my opinion very nice results:

![image](https://user-images.githubusercontent.com/1334853/140619059-30cc072d-c25f-409f-a6c0-6783d91bdc44.png)

By coincidence this matches the look of the Cardinal host panels, so I think I am going with this.
Helps to make Cardinal more distinct and unique too.

**Comment by falkTX** [comment]
Dark mode for the win! :rocket: :rocket: :rocket: 

![image](https://user-images.githubusercontent.com/1334853/140642610-1440ccb3-32d0-4608-abb3-d7b0a7a418f3.png)

Try the latest builds and let me know if it looks good for you, or if this is too much.
Thanks

**Comment by dromer** [comment]
This is for all module packs that have a CC NC license?

**Comment by falkTX** [comment]
> This is for all module packs that have a CC ND license?

For everything. So far I have not had the need to edit a single svg file.
Not all of the files are CC-ND and have restrictions on derivatives, it just makes sense to use the same approach for everything.

**Comment by dromer** [comment]
> it just makes sense to use the same approach for everything.

Not if it 1)  makes it ugly 2) makes it harder to discern modules 3) other reasons

**Comment by vcvrackideas** [comment]
Mutables looks indeed nice for me! And Fundamental too. No surprises - it was a long thread in VCV community on Dark mode and all things on rights and afaik it was not allowed to change them and post on the internet. But since you have not changed SVGs at all is of course your app done it so it looks pretty nice and ok :-)

**Comment by falkTX** [comment]
> > it just makes sense to use the same approach for everything.
> 
> Not if it 1) makes it ugly 2) makes it harder to discern modules 3) other reasons

I was referring to modifying the svg files vs applying changes on the rendered shapes/colors.
Makes sense to not modify any files, since some we are kinda not allowed to anyway, this way the method to give dark look is the same for all.

For the hard to discern argument, not sure if that is really a new problem. A lot of them are very alike already.
I personally think having some modules with dark background (for example ValleyAudio) while at the same time some are very bright (the built-in/core VCV/Fundamental) is a bigger problem.
A kinda consistent look across the modules is great to see, and likely easier on the eyes as well.

**Comment by falkTX** [comment]
@AriaSalvatrice would you be ok with me adding your modules to Cardinal?
Since you are no longer involved in VCV or updating your modules, I thought I should ask first. They seem quite useful for sequencing.
IMO there is no need to change the graphics, but if I some point I do I will make sure to remove your signature.

**Comment by falkTX** [comment]
Breaking news! On the sense that the updates I just push will break existing projects 😂 

I made the previously "Cardinal" plugin the full variant of the plugin, the one that has Audio, MIDI and CV.
What was before just "Cardinal" is now "CardinalFX" (stereo audio in+out, plus MIDI because why not).
The "CardinalSynth" variant remains unaffected.

This change is so that we can expose CV ports to hosts and systems that understand them.
So if for example you are using JACK, this is what you see on your patchbay:

<img width="1280" alt="image" src="https://user-images.githubusercontent.com/1334853/140995589-f05a3327-5dc8-463c-98d1-b1998079c7fb.png">

A new module called HostCV has been added which can and write from these ports, all monophonic yes, but you can easily merge them to send those to poly modules.
For example like this:

<img width="1280" alt="image" src="https://user-images.githubusercontent.com/1334853/140995798-321a757e-b2c4-43b0-9390-d8e980b55b08.png">

The HostCV module has options through right-click that allow to set the CV as bipolar, in case the regular 0-10 range is not something you want.

<img width="200" alt="image" src="https://user-images.githubusercontent.com/1334853/140995991-cd794047-5eed-4692-9ae8-25071f1f0333.png">

That makes it handy to use directly as oct/volt

Obviously this module wont work if loaded on CardinalFX or CardinalSynth, since those do not expose CV ports to the host.
The HostParameters module is unaffected.

**Comment by dromer** [comment]
This is awesome!

First time Rack can actually talk full CV i/o to the outside?

**Comment by falkTX** [comment]
> This is awesome!
> 
> First time Rack can actually talk full CV i/o to the outside?

Weren't there some tools/forks before to have VCV on hardware?
They were very specialized though, this is more generic for supporting JACK and LV2, so we can go all modular.

Specially now that PipeWire is becoming a thing on Linux, we have full Patchbay + CV capabilities mixed with the regular desktop stuff. We can in theory even send and receive CV from the browsers through PipeWire.

One cool idea pops up here.. (will be more doable with patch export, coming soon)
We can have multiple instances of Cardinal connected together (or running in parallel) through JACK/PipeWire, feeding into each other. One can be the "conductor" sending a cv sequence into other instances.
Driving hardware should be much easier with this now in place, since writing jack clients is much easier than a whole VCV module. Together with something like https://github.com/moddevices/mod-spi2jack it could already work on a few existing systems 😱 

VST has nothing like this hah, the modularity of Linux pays off in the end 😎 


PS: I even completely forgot to mention before because it is not 100% finalized yet, but Carla is a module inside Cardinal now too.
So we can load more than just the simple LV2 fx that Ildaeil has on its list.
I need to do a CV2MIDI plugin so this is more useful, but for now at least for FX and random LV2 CV things already works.
It is not 100% finalized because this is a smaller Carla build with less features, plus requires Carla be installed on the system. This last thing makes it so that it does not work on Windows yet, I will have to think if shipping with Carla binaries inside Cardinal Windows builds makes sense or not...
At least it should work on Linux and macOS, assuming you have Carla (as application) installed.

<img width="1280" alt="image" src="https://user-images.githubusercontent.com/1334853/141004064-e0da85be-fb16-4304-b5ff-0e4e59106753.png">

**Comment by dromer** [comment]
Jup, connecting cardinals together was also my idea.

I've done this before using https://github.com/Skrylar/skjack-vcv however this was using audio ports. Having the ability to send CV is even better!

**Comment by Torinbob** [comment]
> First time Rack can actually talk full CV i/o to the outside?

There's no difference between audio and CV in Rack. The regular Rack Audio-8 and Audio-16 modules can send and receive CV.

**Comment by dromer** [comment]
You'd want to connect this to a DC-coupled interface though. And having the ability to talk software-cv in jack together with other jack app sis a win too.
Being able to distinguish between audio and cv signals is quite useful.

**Comment by mxmilkiib** [comment]
For anyone unfamiliar, the metadata systems of JACK and LV2 allow differentiation between audio and CV ports. Further info on https://linuxmusicians.com/viewtopic.php?t=20701

**Comment by Torinbob** [comment]
While Jack and LV2 may allow that distinction, the fact that Rack does not make one is actually a big advantage and makes it very flexible. By introducing a distinction, you introduce limitations -  you make it so you can only send CV into this jack and not audio (or vice versa). Whereas in Rack, you want to send CV into this jack? that's fine. Or audio? that's fine too.

**Comment by falkTX** [comment]
> While Jack and LV2 may allow that distinction, the fact that Rack does not make one is actually a big advantage and makes it very flexible. By introducing a distinction, you introduce limitations - you make it so you can only send CV into this jack and not audio (or vice versa). Whereas in Rack, you want to send CV into this jack? that's fine. Or audio? that's fine too.

Have you ever used JACK? It does not seem like you know what you are talking about, sorry.
For rack it makes sense to have both be the same, but with regular audio/desktop applications and DAWs it does not. Feeding CV signals into a speaker/monitor can literally blow it up.

The distinction inside JACK is purely informational, a tag applied to audio ports (because they **are** audio ports). A similar "hack" can be done for placing OSC messages inside MIDI ports, by abusing the first byte and making it char '#' for bundles and '/' for regular messages. JACK will happily pass the data buffers around, it does not really care what is inside.
This way audio and CV works just fine.

If one really needs to use CV as audio ports or vice versa, this is easy to do by just using a tool that does not differentiate between the 2. Carla does as a precaution for the user, but `jack_connect` for example does not.
Case in point, I tried sending audio through the HostCV module and it works fine. The values are -5/+5 so I have to apply gain reduction to make it scale down to -1/+1 but the audio stream is lossless.

IMO we get the best of both worlds. There is a distinction as a mistake-prevention mechanism but if you know what you are doing you can still go at it treating everything as float buffer stream.

**Comment by falkTX** [comment]
We finally have the file->export function working.
This was a tough one, specially since there was no way to do it on Linux.
The sofd mini-lib used by DPF to handle file browser dialogs on X11 only provides a file-open method, there is no file-save due to lack of text/line input (which is seriously difficult to get right).

Having things working cross-platform is a must for me, so I had to look for solution.
We could go the typical fork+exec for zenity/kdialog etc on Linux, but I always found this approach super ugly and unreliable. These are external tools that can change at anytime or fail to load for arbitrary reasons. Plus fork/exec is not a very good thing to do in plugins in general...
One alternative would be to try and get the "desktop portal" stuff to work, see https://flatpak.github.io/xdg-desktop-portal/portal-docs.html for reference documentation.
The main issue with this last one is that we need DBus, and libdbus has some seriously bad API design (even the authors state you do should not to use it but go with glib or qt high-level APIs instead). Introducing glib as a dependency is a big no, so I accepted the pain and just went to learn libdbus a bit.

libdbus is for sure a very awkward/weird and hard API to use, but after you start to understand the different types and how to create arrays, dicts etc the pain is not so strong anymore.

Skip a few hours and I had some initial implementation working. Some more tests here and there and now it is in a state I feel comfortable shipping with. See https://github.com/DISTRHO/DPF/blob/315c10e1bc80fd43ee32131fe7180bba811604dc/distrho/extra/FileBrowserDialog.cpp#L348 for the result.
So we now have linux desktop native file dialogs called through the plugin. The sofd approach is still there as fallback in case the current desktop-environment does not support the desktop portal dbus methods.

Example of dbus file browser in KDE

![image](https://user-images.githubusercontent.com/1334853/141555638-6370ffdb-7771-4bd7-8e74-29371e522d61.png)

So now with this in place the rest was much easier, finally I could add a file-browser-save action for x11/linux.
Doing the same for macOS and Windows is pretty straight forward, just a couple of lines changes to use "save" instead of "open".

Apart from this, Atelier module was added (by request) and also mscHack.
Because a few modules need some tweaks I setup an org for hosting them https://github.com/CardinalModules outside of DISTRHO project as otherwise it would get filled with too many forks.
I am pushing all the fixes upstream as PRs, I dont think most will get merged but it is good to try anyway.

To make things more fun, as obviously there are not enough blank panels, I started my own 😅 
Since most of the artwork on VCV modules is CC-NC I just went with some Miku artwork which is CC-NC as well, see https://piapro.net/intl/en_for_creators.html

![image](https://user-images.githubusercontent.com/1334853/141569370-e18b5737-6bbf-422c-b522-1ab0b6fe45bb.png)

Fun is part of life too :P and there is not enough Miku in the world anyway

**Comment by dromer** [comment]
yay @ export! finally an easy way to exchange patches between sessions :D 

Having Atelier means we can at least get MI Plaits engine available. Perhaps we can blacklist the Audible MacroOsc2 until we know how to debug https://github.com/DISTRHO/Cardinal/issues/27

**Comment by falkTX** [comment]
Running the debug build through valgrind might be helpful again, since I believe we got rid of the race-conditions in the engine by now (by making it always single-threaded).

Should we setup a ticket for people to share patches suitable for inclusion in the project?
If we start having a good enough collection I think it would make sense to ship with them and have them listed as file->open-demo menu option

**Comment by falkTX** [comment]
Things have slowed down a bit in terms of new stuff. But I am keeping Cardinal source up-to-date/in-sync with Rack v2 branch.
Gave a little update to all plugins just now, a few needed fixes which have been submitted upstream, some still to be merged.
Any needed intermediary fork is placed in https://github.com/CardinalModules

The new fundamental modules look nice in real dark mode :crescent_moon:

![Screenshot_20211123_232858](https://user-images.githubusercontent.com/1334853/143145266-52257d1f-96b3-456c-aa8f-42511fa32b20.png)

**Comment by AriaSalvatrice** [comment]
> EDIT: Actually the GPL-v3.0-only is a problem, because the whole thing is built statically it would make the final binary GPL-v3.0-only, that would be something to avoid

I've been away from code stuff a lot lately and ultimately couldn't really make time to help out on Cardinal, but following a private conversation we had to day, I relicensed the code to GPL3-or-later. Since Cardinal does not allow online modules, it will not bundle Arcane, which was the only module of mine to use CC-BY-NC imagery. I believe that usage of my signature is covered by my own legal guidelines and section 7 of the GPL: as it aims to be a good faith port of my code as I made it, the signature can stay. Please ping me if there's any further legal issue to iron out.

**Comment by falkTX** [comment]
Bringing some news.

I have written a proper readme and also a few project documentation related pages - license page (still need to write about artwork and CC), project overview page, and rack pro vs cardinal differences page.
I was waiting until official Rack v2 was released before doing this part, and with that out of the day, let's go :rocket: :rocket: 
Hopefully newcomers are no longer confused about the whole thing.
Let me know what you think of the readme and linked pages within it.

Additionally a few modules were added:
- Aria Salvatrice (without Arcane and related things; not really finalized tbh, so use with care)
- Bacon Music
- Fehler Fabrik
- Mog
- repelzen
- Sonus Modular

I have been requesting authors to have a GPLv3+ compatible licenses where required.
Happy to report that pretty much all of them have been positive about it, only Atelier and ZZC modules remaining. In the case they say no, these will eventually be removed from Cardinal.

**Comment by zezic** [comment]
@falkTX Is there a possibility for modules to be included in Cardinal on some exceptional rules but without changing the modules license from `GPL-3.0-only` to `GPL-3.0-or-later`?

**Comment by falkTX** [comment]
> @falkTX Is there a possibility for modules to be included in Cardinal on some exceptional rules but without changing the modules license from `GPL-3.0-only` to `GPL-3.0-or-later`?

No, because I don't want to have to remove them later on and having users end up with broken projects.
At some point soon the included modules need to be frozen, so then nothing is removed, only added.

I do not want to deal with the nightmare of (future) GPLv4+ released plugins being incompatible with Cardinal because 1 or 2 modules insisted on being GPLv3-only.
It is still far away from now, but that is no reason for improper planning.

**Comment by jaffasplaffa** [comment]
Just wanted to leave a note about my results in making Cardinal work in Logic. I got it working now on my Mac M1 Pro in Logic 10.7.1. I am running Logic in Silicon mode NOT Intel/Rosetta2 mode.

I use the VST/AU host called Minihost from Image Line, which is free, to load the vst in Logic. You can download it from here, you need to create an account, though:
https://forum.image-line.com/viewtopic.php?f=1919&t=123031?_ga=2.229505502.375400101.1638826766-1760077144.1638826766

Minihost needed to scan my plugin folders and some plugins caused Minihost to crash, so I had to remove some of the plugins from VST folders while scanning, but once I got the plugins scanned it works :) 

<img width="1221" alt="Screenshot 2021-12-06 at 23 51 47" src="https://user-images.githubusercontent.com/14827192/144935717-0c7a52d5-c21b-4f25-9a79-a8320b3fdd42.png">

**Comment by falkTX** [comment]
Very nice! And yeah, the binaries are already M1 compatible.

So Logic doesn't support VST2 or VST3 at all? After all this time, that seems a bit weird..

**Comment by jaffasplaffa** [comment]
No, Logic does not support any VST format, only Audio Units. That's why I had to use the Minihost plugin format "converter" to be able to load the VST in Logic. 

And yes I think also VST's should be added to Logic, but yeah, that's probably a big task to make Apple add that, they stick to their own format Audio Units. it seems.

The Minihost plugin is not M1 native, so I am assuming it's converted via Rosetta2 layer and it actually uses the Intel version of Cardinal. Still new to M1 but first I thought I had to run Logic in Rosetta2 mode, to be able to run Intel plugins, but it seems like even in Logic Silicon mode, the intel and the Silicon plugins works fine and can be used together :)

**Comment by jaffasplaffa** [comment]
Oh...... I just tried reopening the Logic and it crashed the Logic project. I opened another Logic project and added a few instances of Cardinal/Minihost setup and it seems to work even after closing and reopening Logic.

But it seems like there are some midi note timing issues. For example, I have to have a track in Logic highlighted for the midi note timing to be correct. If not selected timing becomes a bit sloppy. 

Anyway, would still be lovely with an actual Audio Unit version, so we don't have to jump through hoops. I am thinking also troubleshooting is easier without the Minihost conversion step involved.

Will do some more testing soon ;)

**Comment by falkTX** [comment]
The MIDI part is not yet 100% jitter free, known issue.
For Cardinal we will have to override the Rack original MIDI driver code so it runs in sync with the audio thread.
This is a side-effect from Rack being standalone first, and then a plugin. It uses extra threads for processing and queuing MIDI events, which has these side effects.

It should be possible to make a stupid silly AU plugin in JUCE that wraps around the Cardinal VST2 plugin, something I think other members of the community can very likely help with. I can outline some details on how such a thing would be possible, if anyone is interested on working on it.

**Comment by jaffasplaffa** [comment]
Ahh didn't know about the midi timing.

Yeah that would be lovely, with the Juce plugin wrapper.I do know a little bit of Juce but not enough to take on such a task.

Anyway, will keep an eye on what's happening, but it's nice to know it's possible to use it in Logic.

**Comment by jaffasplaffa** [comment]
Okay, so today I tried opening the Logic project with a few Cardinal loaded via Minihost again, it worked yesterday and today it just crashes Logic.

**Comment by falkTX** [comment]
Must be one of the modules crashing when loading.
Tricky part is finding out which one...

**Comment by jaffasplaffa** [comment]
Yeah, I don't get a crash report, I just get:

"An Audio Unit plug-in reported a problem which might cause the system to become unstable. Please quit and restart Logic Pro."

So yeah, really hard to say what it might be.

**Comment by falkTX** [comment]
It has been a while since I reported news.

Focus has been on bugfixing and remaining features, with a couple more modules added.
The 2 big points worth mentioning would be that copy&paste now works, and all the file save&open menus have been implemented (with cardinal's own API for this).

As a bonus, in preparation for something to be coming at a later point, we have an integrated built-in text editor now.

![Screenshot_20211218_230233](https://user-images.githubusercontent.com/1334853/146657741-bcc81464-81fe-4522-a4ad-a51385fdb8aa.png)

With all the recent work I would say Cardinal is in beta-state quality at the moment, slowly approaching a release.
There are still a few known issues, that is why it is not in RC-like state yet.
Some of these issues are plugin/module related things, but we need to fix them anyway.

Probably just a few more weeks and all should be fixed up and good to go.
That said, if you look past the currently opened Cardinal issues it is already usable as-is.
All the critical pieces are in place already.

I need to stop the work on Cardinal for the moment to focus on a few other things (there is a huge list of JACK patches to go through, and VST3 support from DPF's side to take care of, among others).
Real-world testing is what Cardinal needs at the moment, so a little time to have fun with it instead of changing code will be good anyway.

Happy holidays. :partying_face:

**Comment by dromer** [comment]
Thnx for the update, wonder what that text editor will do ..

Since the beginning I've been pairing Cardinal up with BespokeSynth, lots of fun!
Looking forward to some better integration (still need to have them fix the exposed parameters issue as well).

**Comment by MarcBoule** [comment]
> wonder what that text editor will do

Hope it's for algorave live-coding :-)

**Comment by falkTX** [comment]
GIving a little update.

I have been tracking down all resource files and their license.
Now nicely filled in the table at https://github.com/DISTRHO/Cardinal/blob/main/doc/LICENSES.md#artwork--panel-licenses
That helps us to see the bigger picture, and clear a bit the situation on licensing.

The CI builds now include modduo, modduox and moddwarf targets, which will make special headless LV2 builds of Cardinal for use in [MOD](https://moddevices.com/)
All the non-headless builds now have OSC enabled, through liblo, as a way to push to those embed devices.

More on CI setup, [PawPaw](https://github.com/DISTRHO/PawPaw) is now used in order to have a few static libs precompiled for us. Which allow us to have libsndfile available during the build, specially handy for audio file handling.
So Ildaeil, making use of Carla's internal audiofile plugin, can now load audio files and have them play in sync with the host, with the disk-streaming if files longer >30s and all its goodness.

![Screenshot_20220120_101229](https://user-images.githubusercontent.com/1334853/150318714-7fccc978-ab40-4172-8e04-9923f200e200.png)

Such functionality is a bit hidden though, so I started a dedicated audio file module, which is basically the same thing that runs in Ildaeil but with a dedicated panel plus list of files in the current folder.
It is still work-in-progress.

![Screenshot_20220120_101336](https://user-images.githubusercontent.com/1334853/150318998-fcd668d2-d8ea-46c3-81fc-6875d1b8347e.png)

There were all sorts of little changes and fixes, a few modules added in too. But nothing too substantial.

There is still the use of CC-ND that Cardinal tries to bypass by loading the files and applying specific drawing rules for each one on the fly, with the idea that CC-ND is not broken if no actual new work is being produced (the "black" version is never saved on disk, and what is in memory is not an svg file but drawing instructions)
While thinking about this, and potentially looking to at least replace the Core panels, I decided to instead just not use them at all.
As-is, the Core panels do not make a lot of sense in a plugin workflow. The audio and midi panels for example, have a list of devices to pick from but we will only be able to pick the "Cardinal" plugin driver. Makes that whole part kinda pointless.
Not to mention, the audio module does all sorts of tricks in order to get multiple audio devices in sync, with ringbuffers, sample rate converters, and even more buffers.
It is really wasteful when we get the audio from a single source, with the sample rate already set.
So I began "host audio" cardinal-specific module, bypassing all the driver/device handling from Rack and just using the buffers directly.

Just so you get an idea, this is the original code from Rack: https://github.com/VCVRack/Rack/blob/v2/src/core/Audio.cpp
And this the new code from cardinal: https://github.com/DISTRHO/Cardinal/blob/main/plugins/Cardinal/src/HostAudio.cpp
In theory this should consume less CPU, but didn't measure it.

With this change, the "old" core audio modules do nothing now, I completely removed the audio driver handling. Later on likely will do the same for MIDI.
The panel design is still work in progress, but already works: (2ch on left, 8ch on right)

![Screenshot_20220120_102259](https://user-images.githubusercontent.com/1334853/150320773-a584c4ab-bd2b-4600-aef4-de082a2e62f5.png)

To finalize, I am setting a deadline to finish all this - 14 February.
Otherwise we never have a release. Still 25 days ago, seems enough time...

**Comment by dromer** [comment]
For all the core audio and midi modules it makes sense to redesign, since Cardinal doesn't need to have device selection and configuration here. Although possibly in the future >1 midi ports could be useful (which plugin formats and hosts support that? has some use-cases for sure).

**Comment by falkTX** [comment]
> Although possibly in the future >1 midi ports could be useful (which plugin formats and hosts support that? has some use-cases for sure).

considering each rack cable can only carry through 16 voices, I think this idea is quite pointless. (as much as I would like it to be differently)
you can easily use multiple midi channels for a total of 16 parts of 16 max polyphony each.

Cardinal being a plugin means you can use multiple instances, and that is the whole point anyway.
If the 16x16 voices are not enough, just run another instance of cardinal in another daw track.

**Comment by dromer** [comment]
true, true. multi-instance also has some more preference for me in project management, so better focus on that together with preset handling.

**Comment by falkTX** [comment]
Quick but very needed news.

I just pushed a change that takes out all the uncertain or not yet resolved license issues for module artwork.
The intention is that the modules we have on next builds should, in theory, be 100% safe to use and never going to be removed.
Removed modules will slowly be added back again once we get the license situation sorted out, or we put them in a denylist for modules we know won't be possible to use (unless authors change their terms).

The list of temporary removed modules is:
- AS
- Atelier
- Audible Instruments
- Befaco
- Bidoo
- cf
- Fundamental
- ihtsyn
- ZZC

ZZC I am quite certain will be added back, as the developer has been quite open to Cardinal. For the sake of impartiality I removed any that had license issues right now, no matter how likely it is for them to be resolved.

**Comment by falkTX** [comment]
The core modules are being slowly replaced by Cardinal counterparts.
Host Audio and Host MIDI modules should be working now, though the Audio2 version is missing the peak meters.

![image](https://user-images.githubusercontent.com/1334853/150783478-89190939-273d-4967-ac53-5b70848884b5.png)

Still WIP and yeah they look boring, but should be good enough for a first release. Getting things moving is the first step.
This is turning into a positive in the end, as many optimizations can be done, also allowing many audio/midi modules at the same time too. You can have many audio2 modules loaded, the audio is combined at the end from all the instances. (the same applies to MIDI output)

![Screenshot_20220124_123607](https://user-images.githubusercontent.com/1334853/150783970-7803a868-0599-45be-9279-4ae2829aecf5.png)

**Comment by falkTX** [comment]
Something small but interesting, we can keep existing background by setting OpenGL clear color as full transparency.
This allows to draw OpenGL stuff on top of the panel background, which can be a gradient drawn via NanoVG.
Result is quite cool, even if for a toy, but somehow looks good and professional (to me at least anyway).

![image](https://user-images.githubusercontent.com/1334853/150885515-e40ddf14-b35d-4b06-80ab-cce4ad3d3777.png)

The text editor module from Cardinal now fully replaces the Core Notes one.
It will detect project state from Notes and adjust size to match what the Notes has (the text editor is resizable in Cardinal).
As a little bonus you can now move the text editor around the Rack by grabbing its top status bar.

The cf modules were added back, I manually fixed up the license issues (2 proprietary fonts were in use, replaced by open licensed ones).

**Comment by falkTX** [comment]
Oh here is the current family of the Cardinal modules so far:

![Screenshot_20220125_003347](https://user-images.githubusercontent.com/1334853/150888327-c319b3f9-5bd2-4965-bdfb-1046d7fb2ebc.png)

Audio2 and AudioFile still WIP on their panels, but functionality-wise are ok.

**Comment by falkTX** [comment]
Stuff for fun

Drawing glBars and audiofile things on layer 1

![image](https://user-images.githubusercontent.com/1334853/150980987-7392b907-4f05-4572-92af-78a3bb0455fe.png)

Embedding outside widgets, like a video player, inside Cardinal

![image](https://user-images.githubusercontent.com/1334853/150980937-fd400aa5-2ca7-4d45-8644-cb310b2649a4.png)
![image](https://user-images.githubusercontent.com/1334853/150980973-bb8a35c2-39fd-4d18-9211-330eb3e3aa2a.png)

**Comment by dwat3r** [comment]
ah now I want a module where I can modulate the video playback like simpliciter :) glitch videos with Cardinal!

**Comment by falkTX** [comment]
The Host MIDI-CC is implemented now, leaving only MIDI-Map and MIDI-Gate remaining for the Rack core reimplementations.
This is the Cardinal module family right now:

![Screenshot_20220129_165700](https://user-images.githubusercontent.com/1334853/151669878-f7fe2585-a113-47ac-b88e-76f15dedbe78.png)

They look so happy together :star2: 

The audio2 module now has a custom knob and level meters, so it more closely resembles the original from Rack Code while still being sufficiently different to not be a derivative.

On the MIDI-CC I opted to make the CC grid 3x9, as I think channel pressure and pitchbend are quite useful to have mapped to CV, and they were a bit cumbersome on the original Rack modules to use (channel pressure is not supported there, and pitchbend requires enabling MPE). I didn't actually implement these last 2, but should be quite simple.

Not sure what to do about the (Rack core) Blank module, it is nice for being resizable, but dont want any potential issues of "visual design" being under copyright, even though it is all drawn with code.
I have written to Andrew a couple of days ago about Cardinal and related things, but got no reply yet. I will remove the VCV Blank module if I still have no reply after finishing the MIDI-Map and MIDI-Gate.

Related to modules and licenses, Bidoo has confirmed being okay with our runtime dark mode tricks (in fact they are working on an official dark mode right now, so maybe eventually Cardinal runtime tricks can be removed). Emilie from Mutable Instruments also gave green light for Cardinal dark mode mode of the VCV-made Audible Instruments.
AS author doesnt seem to understand opensource, or at least I got that impression, and it is unclear at the moment if they will even keep their modules source code public or just go proprietary.

In any case, we can always look for modules with better and clearer license/usage terms.

Seems to me everything still on track for a first Cardinal release in 16 days. Soon soon.

**Comment by falkTX** [comment]
And the core replacement is done, the family is now complete.

![Screenshot_20220130_005020](https://user-images.githubusercontent.com/1334853/151682629-cd12188b-445d-41a5-b14d-9df719628404.png)

Still no answer from Andrew, so Rack Core Blank panel is gone. If you load a patch from VCVRack that used it, it gets replaced with Miku. A fair trade :dancers:

**Comment by falkTX** [comment]
The super experimental embed widget / video player through MPV works a tiny bit better now.
Turns out we can use `XShape` to hide specific parts of a Window, this can be quite useful here. I made it so that the menubar area is cropped in the widget side, and everything is hidden away if the module browser is open.
That makes it a bit usable now.

https://user-images.githubusercontent.com/1334853/151729685-badb3d7e-918c-4d98-b346-8d30249daf83.mp4

**Comment by falkTX** [comment]
FYI Just did some changes regarding cable values behaviour.

I was not aware before, for not having yet dig deep enough into the code, but every cable introduces 1 sample of latency in Rack because of how the engine works.
The engine runs on every frame, but ports with connections require the next cycle for the target port to see new values.
The cables basically accumulate and slowly add in bits of latency everywhere..
This is quite annoying, to be quite honest I expected there to be no latency on straight connections like A->B->C, disappointed that it is not the case...

But wait! Now it is!
Or at least it starts so. I made it so that (output) ports store their current connections/cables. Every time a module port calls `setVoltage` the other (input port) side gets its value updated at that very moment, no need to wait for the next cycle.
This makes the A->B->C case have zero latency!

Now this is great but there is still more to do. Carla and Ildaeil (plugin hosts) have their own buffering because running regular audio plugins at 1 frame is just unfeasible. I am thinking of making them have the same exact buffer size as Cardinal in the DAW, so in theory they would run in sync with everything else and also have 0 latency, but need to do some tests to make sure of that.

Another issue still remaining is the lack of a "graph" context in Rack. Rack's engine simply goes through all modules one by one in the order they were created. So if we add say A, B, C modules and connect C->B->A the zero-latency cable stuff doesnt help at all. In order to have reduced latency the modules would need to be processed in order of their connections, which is what a "graph" entails. Everytime a new connection is made the graph would be rechecked for the most optimal path of processing. Basically how any modular host/patchbay works.
I am not familiar on creating such graphs so won't do this myself, I only know they are quite complex to code, so it is understandable that Rack doesnt have one yet.

In other news, AudioFile and Ildaeil panels got a little reorganized.
Slightly related, I updated the imgui stuff and made it recreate the GL stuff on zoom level changes so it always looks crisp.

![image](https://user-images.githubusercontent.com/1334853/152663740-86b23268-e72c-4114-9ded-25eaeb791fae.png)


Also started playing with expanders. The stuff from Ildaeil was split off to create the midi-to-host modules expander.
Wanted to make sure to signal that the connection was active, so when close to a module it works with, the expander will expand its port background.

![Screenshot_20220206_003946](https://user-images.githubusercontent.com/1334853/152663790-d8483e6d-ac92-4a8d-8f7e-c65bb107342b.png)

The background color is wrong though, on Cardinal the white layer is supposed to be for outputs, so I will still change before release.

And speaking of releases, it is just 8 days away now :)

**Comment by zezic** [comment]
@falkTX That's all nice, but how do you plan to handle the circular feedback loops? It will need some type of graph clustering to be able to process the A->B->C chains if they have a C->A connection?

Ability to extract the non-feedbacking chains and run them with 0 cable latency is good to have, but how to make the patch perform consistently when rack operator makes changes is a problem.

Maybe it would help to access the archive of the Rack GitHub issues and also to search the community forum for related discussions, it can fire some light on the question.

**Comment by falkTX** [comment]
> @falkTX That's all nice, but how do you plan to handle the circular feedback loops? It will need some type of graph clustering to be able to process the A->B->C chains if they have a C->A connection.

that is unsolvable unless there is specific dedicated APIs for those.
JACK has a similar problem where you can use feedback loops. IMO it is acceptable to have 1 sample latency for those.
The audio graph as implemented in JUCE doesnt even support feedback loops at all, and mutes the output if one is detected.


> Maybe it would help to access the archive of the Rack GitHub issues and also to search the community forum for related discussions, it can fire some light on the question.

There is no archive anywhere that I could find, if you know of a platform that has a copy please tell.
Joining the forums is something I am undecided about. It could be seen as trying to promote an unofficial product and generate drama/anger/etc.. And anyway discourse forums are not the place for such talks. The Rack GitHub issues should have never been closed in the first place! I dont feel comfortable start a discussion around something where we already have precedence of community being silenced. The same way the GH issues were closed, other topics can be closed just as well.

**Comment by falkTX** [comment]
MIDI output expander is implemented, so plugins loaded in Carla or Ildaeil that generate MIDI out will now be able to be part of the project too.
Did an experiment with Carla's internal midi file player, which gets converted to CV, back to MIDI for an instance of zynaddsubfx. Works fine :tada: 
![image](https://user-images.githubusercontent.com/1334853/152694507-76833e7e-1e4c-49bf-87dd-6508a98217c7.png)

**Comment by zezic** [comment]
> IMO it is acceptable to have 1 sample latency for those.

I mean, sure, that's absolutely acceptable, but for me it feels that it will be a pretty complex algorithm to find 0-sample-latency segments, cluster them and decide where to use the 1-sample-delay cables.

> There is no archive anywhere that I could find, if you know of a platform that has a copy please tell.

Sadly, I can't find it either.

> Joining the forums is something I am undecided about.

Ah, I mean, you can use them read-only. However, after quick search I can't find any discussions about cable delays and graphs which would be productive enough. But it's an interesting topic.

**Comment by falkTX** [comment]
> > IMO it is acceptable to have 1 sample latency for those.
> 
> I mean, sure, that's absolutely acceptable, but for me it feels that it will be a pretty complex algorithm to find 0-sample-latency segments, cluster them and decide where to use the 1-sample-delay cables.

It is not *that* complex, there are even several known algorithms for it.
JACK basically works this way, when you make new connections the entire graph is checked for the optimal processing path. If a new path is found a graph reorder callback is triggered.

Maybe with Rack things are a bit harder because the amount of connections will be bigger, but still doable.
I have seen people with setups with a big amount of connections while using JACK, and it still behaves as you expect, e.g. https://kx.studio/screenshots/news/carla-2.2_usage.png



> > Joining the forums is something I am undecided about.
> 
> Ah, I mean, you can use them read-only. However, after quick search I can't find any discussions about cable delays and graphs which would be productive enough. But it's an interesting topic.

Thanks for looking! And sure, I already do see the topics there from time to time.
But serious dev talk doesnt seem to happen much on their forums.

**Comment by falkTX** [comment]
It is release day!
Wrote a few details about Cardinal on my KXStudio page https://kx.studio/News/?action=view&url=cardinal-2202-is-now-released

The releases page https://github.com/DISTRHO/Cardinal/releases has the usual builds attached, they should downloadable by anyone, even those without a github account.

**Comment by falkTX** [comment]
The very first piece of media after the tagged release,
From our dear unfa :sunglasses: 
https://www.youtube.com/watch?v=HapldC7rROc

**Comment by falkTX** [comment]
I didnt hear anything incorrect.
The crash was weird, maybe Cardinal inside Cardinal, but I have tested that to work fine before.

Thanks for the quick video!

**Comment by unfa** [comment]
You're welcome! I was hoping maybe I can manage to do this yesterday, but I wouldn't have the demo song to show, and I think it was worth the delay.

Are there any debugging tips? Maybe I could collect some logs or coredumps after such crashes to help fix the issues?

**Comment by falkTX** [comment]
> Are there any debugging tips? Maybe I could collect some logs or coredumps after such crashes to help fix the issues?

The release binaries are quite useless for debugging, as everything is optimized and stripped for the best performance.
There are debug builds on the "nightly" github actions though, which run quite poorly but allow for full backtrace and debugging.

But I think the best for helping to fix issues is just to get a nice set of reproducible steps more than anything else.

**Comment by falkTX** [comment]
To keep you all updated on what is going on. Quite a few modules were added thanks to @dromer, with me just helping along the way. We finally have Befaco :rocket: 
For now it is best to not add any more, focus goes on stabilizing the ones we have right now.

I noticed a macOS issue where the disabling of "update rate limit" breaks a few hosts. Situation is a little weird/unusual.
Some hosts always repaint the plugin UIs even if it has nothing new to draw, that makes Cardinal use a lot of CPU wastefully - this includes Ableton Live, Carla and REAPER; some other hosts only repaint as necessary - including FLStudio and Renoise.
The way original Rack handles this is just to do a small sleep if it detects repaints happening too often. I dont think this is a good option, specially in a plugin, as it will slow down other things the daw does.
So the "update rate limit" option is there for macOS now, but note it doesn't work on all hosts.

Current plan is to fix up issues related to the new added modules, then a couple of other ones reported, and also do a windows installer. As much as I think copying some folders around is okay, seems not everyone agrees... :/
The individual plugin bundles will remain as release binaries, but we will have a windows installer option too.

There is no specific date set, but target is to do a 22.03 release this month with the newly added stuff and fixes.

PS: unfa is doing a livestream using Cardinal very soon, in about 10mins from posting this message! https://www.youtube.com/watch?v=Tpd_vZJwGNQ1 :rocket:  :rocket:  :rocket:

**Comment by falkTX** [comment]
On something maybe expected, maybe unexpected, I have been testing leveraging JUCE as a way to get a few more plugin formats working. Basically bridging between DPF and JUCE, giving a few more options to DPF plugins.
It is not something that belongs in DPF, because licenses, but can be quite useful in selected cases which Cardinal is one of them.

An obvious target is to quickly get an AU plugin without having to do the full implementation on DPF side, as a few DAWs like ProTools only support AU. I dont have a ProTools license, but do have GarageBand that is in a similar situation only support AUs.
Does it work? ..yes :)

![image](https://user-images.githubusercontent.com/1334853/157512935-ae2e21e3-0f70-4409-ac2f-d4677de6df96.png)

Got audio, transport information and UI working. Parameters, MIDI and state save/load not yet. But first was just checking if this was doable anyway.
I am not sure if this will be part of 22.03 release, since it is very new and I am still doing tests, but it is great to have as an option.

**Comment by falkTX** [comment]
Some news!
So the AU stuff was finalized, at least for the use I wanted to make of it. I do not use it myself, but give it some testing to ensure all required functionality is present. Testing is welcome.
(on the technical side, this is pretty much just leveraging JUCE as a middle layer between DPF and AU. Initial testing and development was done on Linux with VST3 and Standalone formats, and then double-check on macOS with AU. Since it is all JUCE stuff, I suspect it works fine).
Only issue I saw was resizing being a bit jerky, have some ideas for how to make it better, will try them later.

On the Windows side, happy times, we got an installer now!
The resources are stored on a single place, reducing wasted space. And yeah it puts the plugin files on the correct places, no manual copying necessary.
The situation with Carla (the plugin host) has been solved on Windows too. Now Cardinal builds the whole GUI toolchain needed (Qt, Python and PyQt) and makes it an option during install.

![image](https://user-images.githubusercontent.com/1334853/158881331-3318b3fe-f0f0-4af7-b071-bf6b349a5bb8.png)

Carla is running! :rocket: 

![Screenshot_20220317_194657](https://user-images.githubusercontent.com/1334853/158883570-f236a464-a954-4329-aeb2-136482e65982.png)

And speaking of Windows, most of the issues present on the 32bit build were fixed. Some still remain, but way less than before. This will be an on-going effort, as VCV Rack never supported a 32bit OS so we are on our own for this.

With these changes and the added modules already make it worth to have a new release.
I have set the date for that to be 21th of March, the start of Spring season. Will do a few more fixes as possible until then.

**Comment by falkTX** [comment]
Spring is here, and so is a new release.
Details at https://kx.studio/News/?action=view&url=cardinal-2203-released

Time to party :tada:

**Comment by wpostma** [comment]
I'm a Visual C++ dev not much MSYS2 experience but I have a high dpi Windows DAW workstation and would be willing to help with the current high dpi issues on various windows vst hosts (bitwig, reaper, ardour) with cardinal .

**Comment by falkTX** [comment]
I dont use msys2, that is why :)
The windows builds are cross-compiled from linux, it has been many years since I actually used windows as an OS.

That said, https://github.com/lv2/pugl is the underlying layer used by DPF (in turn used by Cardinal). So a first step would be to try to reproduce the issue with it.

**Comment by falkTX** [comment]
Just tagged 22.04 :tada: 
https://github.com/DISTRHO/Cardinal/releases/tag/22.04

Note: I will stop writing updates here.
Cardinal is more or less an established project now, and known to work well for quite a few people.
If there are issues, you know where to report them.
Some are still unfixed, which likely will be fixed in due time. Not everything is perfect but this is a fully free plugin in the end. Help is always appreciated for those that can provide it, be as it source code fixes or other means.

PS: builds still going, binaries will be up soon.

**Comment by faseh329** [comment]
Download 🟢Link👉👉[FL Studio 2025 Activated Version](https://tinyurl.com/fl-studio-2025-activated)🟢🟢

**Comment by faseh329** [comment]
Download 🟢Link👉👉[FL Studio Activated Version](https://tinyurl.com/fl-studio-2025-activated)🟢🟢

**Comment by faseh329** [comment]
Download 🟢Link👉👉[FL Studio 2025 Cracked Version](https://tinyurl.com/fl-studio-2025-activated)🟢🟢

**Comment by faseh329** [comment]
Download 🟢Link👉👉[FL Studio Reg Key2025 Activated Version](https://tinyurl.com/fl-studio-2025-activated)🟢🟢

**Comment by faseh329** [comment]
Download 🟢Link👉👉[FL Studio Setup Activated Version](https://tinyurl.com/fl-studio-2025-activated)🟢🟢
